src/jit/compile.c

```diff
@@ -88,61 +125,48 @@ MVMJitCode * MVM_jit_compile_graph(MVMThreadContext *tc, MVMJitGraph *jg) {
 
     /* Get the basic block labels */
     code->num_labels = jg->num_labels;
-    code->labels = MVM_malloc(sizeof(void*) * code->num_labels);
+    code->labels = MVM_calloc(code->num_labels, sizeof(void*));
+
     for (i = 0; i < code->num_labels; i++) {
-        MVMint32 offset = dasm_getpclabel(&state, i);
+        MVMint32 offset = dasm_getpclabel(cl, i);
         if (offset < 0)
             MVM_jit_log(tc, "Got negative offset for dynamic label %d\n", i);
         code->labels[i] = memory + offset;
     }
 
-    /* Copy the deopts, inlines, and handlers. Because these use the label index
-     * rather than the direct pointer, no fixup is necessary */
-    code->num_bbs      = jg->num_bbs;
-    code->bb_labels    = COPY_ARRAY(jg->bb_labels, jg->num_bbs, MVMint32);
-
-    code->num_deopts   = jg->num_deopts;
-    code->deopts       = code->num_deopts ? COPY_ARRAY(jg->deopts, jg->num_deopts, MVMJitDeopt) : NULL;
-    code->num_handlers = jg->num_handlers;
-    code->handlers     = code->num_handlers ? COPY_ARRAY(jg->handlers, jg->num_handlers, MVMJitHandler) : NULL;
-    code->num_inlines  = jg->num_inlines;
-    code->inlines      = code->num_inlines ? COPY_ARRAY(jg->inlines, jg->num_inlines, MVMJitInline) : NULL;
-
-    /* clear up the assembler */
-    dasm_free(&state);
-    MVM_free(dasm_globals);
+    /* Copy the deopts, inlines, and handlers. Because these use the
+     * label index rather than the direct pointer, no fixup is
+     * necessary */
+    code->num_deopts   = jg->deopts_num;
+    code->deopts       = code->num_deopts ? COPY_ARRAY(jg->deopts, jg->deopts_num) : NULL;
+    code->num_handlers = jg->handlers_num;
+    code->handlers     = code->num_handlers ? COPY_ARRAY(jg->handlers, jg->handlers_alloc) : NULL;
+    code->num_inlines  = jg->inlines_num;
+    code->inlines      = code->num_inlines ? COPY_ARRAY(jg->inlines, jg->inlines_alloc) : NULL;
 
-    code->seq_nr = MVM_incr(&tc->instance->jit_seq_nr);
+    /* add sequence number */
+    code->seq_nr       = MVM_incr(&tc->instance->jit_seq_nr);
```

Comment:
Since the JIT now runs on the spesh thread only, an atomic increment ain't needed here.

Reply:
Agreed, and i was actually thinking of moving the assignment of the sequence number to spesh graph time. Will do that in some other commit though.